compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -702,10 +667,93 @@ private static BIntersectionType defineImmutableUnionType(Location pos, Types ty
+                                                                                    unionTypeNode, env);
+        typeDefinition.pos = pos;
+        return immutableType;
+    }
+
+    private static BIntersectionType defineImmutableBuiltInUnionType(Location pos, Types types, SymbolEnv env,
+                                                                     PackageID pkgId, BSymbol owner,
+                                                                     SymbolTable symTable,
+                                                                     BLangAnonymousModelHelper anonymousModelHelper,
+                                                                     Names names, Set<BType> unresolvedTypes,
+                                                                     BUnionType type, BType originalType) {
+        BTypeSymbol origBuiltInUnionTypeSymbol = type.tsymbol;
+
+        Optional<BIntersectionType> immutableTypeOptional = Types.getImmutableType(symTable, pkgId, type);
+        if (immutableTypeOptional.isPresent()) {
+            return immutableTypeOptional.get();
+        }
+
+        BUnionType effectiveType;
+        if (type.tag == TypeTags.JSON) {
+            effectiveType = defineImmutableJsonType(env, pkgId, owner, names, (BJSONType) type);
+        } else {
+            effectiveType = defineImmutableAnydataType(env, pkgId, owner, names, (BAnydataType) type);
+        }
+
+        BIntersectionType immutableBuiltInUnionIntersectionType =
+                createImmutableIntersectionType(pkgId, owner, originalType, effectiveType, symTable);
+        Types.addImmutableType(symTable, pkgId, type, immutableBuiltInUnionIntersectionType);
+
+        return handleImmutableUnionType(pos, types, env, pkgId, owner, symTable, anonymousModelHelper, names,
+                                        unresolvedTypes, type, origBuiltInUnionTypeSymbol, type.getMemberTypes());
+    }
+
+    private static BAnydataType defineImmutableAnydataType(SymbolEnv env, PackageID pkgId, BSymbol owner, Names names,
+                                                           BAnydataType type) {
+        BTypeSymbol immutableAnydataTSymbol = getReadonlyTSymbol(names, type.tsymbol, env, pkgId, owner);
+
+        if (immutableAnydataTSymbol != null) {
+            BAnydataType immutableAnydataType =
+                    new BAnydataType(immutableAnydataTSymbol,
+                                     immutableAnydataTSymbol.name, type.flags | Flags.READONLY,
+                                     type.isNullable());
+            immutableAnydataTSymbol.type = immutableAnydataType;
+            return immutableAnydataType;
+        }
+         return new BAnydataType(immutableAnydataTSymbol,
+                                 getImmutableTypeName(names, TypeKind.ANYDATA.typeName()),
+                                 type.flags | Flags.READONLY, type.isNullable());
```

Comment:
With the Codecov warning, I believe the type symbol is correctly set for 'AnyData' and it cannot be null. So Do we need this logic?

Reply:
Shall we create an issue for this? This PR just moves the logic, and I think we need to test different scenarios properly before removing this.

Reply:
Create a general issue to track this https://github.com/ballerina-platform/ballerina-lang/issues/39499